src/rust/engine/engine_cffi/src/lib.rs

```diff
@@ -649,15 +649,18 @@ pub extern "C" fn rule_subgraph_visualize(
   })
 }
 
-// TODO(#4884): This produces "thread panicked while processing panic. aborting." on my linux
-// laptop, requiring me to set RUST_BACKTRACE=1 (or "full") to get the panic message.
 #[no_mangle]
 pub extern "C" fn set_panic_handler() {
   panic::set_hook(Box::new(|panic_info| {
-    let mut panic_str = format!(
-      "panic at '{}'",
-      panic_info.payload().downcast_ref::<&str>().unwrap()
-    );
+    let payload = panic_info.payload();
+
+    let mut panic_str = match payload.downcast_ref::<&str>() {
+      Some(s) => format!("panic at '{}'", s),
+      None => match payload.downcast_ref::<String>() {
+        Some(s) => format!("panic at '{}'", s),
+        None => "Non-string panic payload at unknown location".to_string(),
+      },
+    };
```

Comment:
I didn't try this, but I don't think this will typecheck because it's ambiguous whether the `|s|` in the `map` is of type `&str` or `String`. The double-match seemed like the most straightforward way to make this work given that. The payload is of type `Any` and as far as I can tell the API doesn't give you a way to get the memory address (which I'm not sure would be helpful anyway), or any other interesting information about the value.

Reply:
You can add `.map(|s| s.to_string())` to convert the `&str` to a `String`, I think. I think that this formulation is more readable than the nested `match`, which I prefer for code that runs in a panic handler.

And it looks like the `{:p}` format specifier might address the memory address issue: https://stackoverflow.com/a/27852760/2518889. It would at least tell us whether the payload is a null pointer (unless that's ensured by rust using a `&`?). The default python `str()` function for user-defined classes just prints the class name and the memory address, so there's some precedent for using the address as a default printable representation.